File: clients/libjabi/peripherals/gpio.cpp
#include <libjabi/interfaces/interface.h>

namespace jabi {

#include <jabi/peripherals.h>
#include <jabi/peripherals/gpio.h>

void Device::gpio_set_mode(int idx, GPIODir dir, GPIOPull pull, bool init_val) {
    iface_req_t req = {
        .periph_id   = PERIPH_GPIO_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = GPIO_SET_MODE_ID,
        .payload_len = sizeof(gpio_set_mode_req_t),
        .payload     = {0},
    };

    auto args = reinterpret_cast<gpio_set_mode_req_t*>(req.payload);
    args->direction = static_cast<uint8_t>(dir);
    args->pull = static_cast<uint8_t>(pull);
    args->init_val = static_cast<uint8_t>(init_val);

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

void Device::gpio_write(int idx, bool val) {
    iface_req_t req = {
        .periph_id   = PERIPH_GPIO_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = GPIO_WRITE_ID,
        .payload_len = sizeof(gpio_write_req_t),
        .payload     = {0},
    };

    auto args = reinterpret_cast<gpio_write_req_t*>(req.payload);
    args->val = static_cast<uint8_t>(val);

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

bool Device::gpio_read(int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_GPIO_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = GPIO_READ_ID,
        .payload_len = 0,
        .payload     = {0},
    };

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != sizeof(gpio_read_resp_t)) {
        throw std::runtime_error("unexpected payload length");
    }
    auto ret = reinterpret_cast<gpio_read_resp_t*>(resp.payload);
    return ret->val;
}

};

File: clients/rust/src/peripherals/gpio.rs
use crate::interfaces::InterfaceRequest;
use crate::Error;
use deku::{DekuContainerRead, DekuContainerWrite, DekuRead, DekuUpdate, DekuWrite};

pub enum GPIODir {
    Input = 0,
    Output = 1,
    OpenDrain = 2,
    OpenSource = 3,
}

pub enum GPIOPull {
    None = 0,
    Up = 1,
    Down = 2,
    Both = 3,
}

enum Func {
    SetMode = 0,
    Write = 1,
    Read = 2,
}

fn gen_req(f: Func, idx: usize, payload: Vec<u8>) -> Result<InterfaceRequest, Error> {
    let mut r = InterfaceRequest {
        periph_id: crate::InstID::GPIO as u16,
        periph_idx: idx as u16,
        periph_fn: f as u16,
        payload_len: 0,
        payload: payload,
    };
    r.update().map_or(Err(Error::PacketFormat), |_| Ok(r))
}

impl crate::Device {
    pub fn gpio_set_mode(
        &self,
        idx: usize,
        dir: GPIODir,
        pull: GPIOPull,
        init_val: bool,
    ) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetModeRequest {
            direction: u8,
            pull: u8,
            init_val: u8,
        }
        let req = SetModeRequest {
            direction: dir as u8,
            pull: pull as u8,
            init_val: init_val as u8,
        };
        let resp = self.send(&gen_req(Func::SetMode, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn gpio_write(&self, idx: usize, val: bool) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct WriteRequest {
            val: u8,
        }
        let req = WriteRequest { val: val as u8 };
        let resp = self.send(&gen_req(Func::Write, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn gpio_read(&self, idx: usize) -> Result<bool, Error> {
        #[derive(DekuRead)]
        #[deku(endian = "little")]
        struct ReadResponse {
            val: u8,
        }
        let resp = self.send(&gen_req(Func::Read, idx, vec![])?)?;
        match ReadResponse::from_bytes((&resp, 0)) {
            Ok(((s, _), ret)) if s.len() == 0 => Ok(ret.val != 0),
            _ => Err(Error::PacketFormat),
        }
    }
}
